sources/academy/webscraping/web_scraping_for_beginners/crawling/headless_browser.md

```diff
@@ -18,140 +18,180 @@ A headless browser is simply a browser that runs without a user interface (UI).
+
+:::
+
+If we added an appropriate selector to our original `CheerioCrawler` code, it would not extract the information, but browser automatically fetches and renders this extra data.
+
+Let's add this new extractor to our code. It collects the names of the recommended products.
+
+```js title=browser.js
+// ...
+const recommendedProducts = $('.product-recommendations a.product-item__title')
+    .map((i, el) => $(el).text().trim())
+    .toArray();
+// ...
+await Dataset.pushData({
+    // ...
+    recommendedProducts,
+});
+```
+
+And here's the complete, runnable code:
 
-If you always want the scraper to run headless, it might be better to hardcode it in the source. To do that, we need to access [Playwright's launch options](https://playwright.dev/docs/api/class-browsertype#browser-type-launch-option-headless). In Crawlee we can do that [in the `PlaywrightCrawler` constructor](https://crawlee.dev/api/playwright-crawler/interface/PlaywrightLaunchContext).
+```js title=browser.js
+import { PlaywrightCrawler, Dataset } from 'crawlee';
 
-```js
 const crawler = new PlaywrightCrawler({
-    launchContext: {
-        launchOptions: {
-            headless: true, // setting headless in code
-        },
-    },
-    requestHandler: async ({ page, request }) => {
-        // ...
+    // We removed the headless: false option to hide the browser windows.
+    requestHandler: async ({ parseWithCheerio, request, enqueueLinks }) => {
+        console.log(`Fetching URL: ${request.url}`)
+
+        if (request.label === 'start-url') {
+            await enqueueLinks({
+                selector: 'a.product-item__title',
+            });
+            return;
+        }
+
+        // Fourth, parse the browser's page with Cheerio.
+        const $ = await parseWithCheerio();
+
+        const title = $('h1').text().trim();
+        const vendor = $('a.product-meta__vendor').text().trim();
+        const price = $('span.price').contents()[2].nodeValue;
+        const reviewCount = parseInt($('span.rating__caption').text());
+        const description = $('div[class*="description"] div.rte').text().trim();
+        // We added one more extractor to get all the recommended products.
+        const recommendedProducts = $('.product-recommendations a.product-item__title')
+            .map((i, el) => $(el).text().trim())
+            .toArray();
+
+        await Dataset.pushData({
+            title,
+            vendor,
+            price,
+            reviewCount,
+            description,
+            // And we saved the extracted product names.
+            recommendedProducts,
+        });
     },
-    // ...
 });
+
+await crawler.addRequests([{
+    url: 'https://warehouse-theme-metal.myshopify.com/collections/sales',
+    label: 'start-url',
+}]);
+
+await crawler.run();
```

Comment:
Yeah, I couldn't decide either. This is more explicit whereas putting it to run is neater. In the end I went for this explicit option because without it, there would be no notion of "adding requests to the queue" at all.